include/flang/Optimizer/Dialect/FIROps.td

```diff
@@ -0,0 +1,2779 @@
+// All operations automatically get a prefix of "fir.".
+class fir_Op<string mnemonic, list<OpTrait> traits>
+  : Op<fir_Dialect, mnemonic, traits>;
+
+// Base class for FIR operations that take a single argument
+class fir_SimpleOp<string mnemonic, list<OpTrait> traits>
+  : fir_Op<mnemonic, traits> {
+
+  let assemblyFormat = [{
+    operands attr-dict `:` functional-type(operands, results)
+  }];
+}
+
+// Base builder for allocate operations
+def fir_AllocateOpBuilder : OpBuilder<
+  "Builder *builder, OperationState &result, Type inType,"
+  "ArrayRef<Value> lenParams = {}, ArrayRef<Value> sizes = {},"
+  "ArrayRef<NamedAttribute> attributes = {}",
+  [{
+    result.addTypes(getRefTy(inType));
+    result.addAttribute("in_type", mlir::TypeAttr::get(inType));
+    result.addOperands(sizes);
+    for (auto namedAttr : attributes)
+      result.addAttribute(namedAttr.first, namedAttr.second);
+  }]>;
+
+def fir_NamedAllocateOpBuilder : OpBuilder<
+  "Builder *builder, OperationState &result, Type inType, StringRef name,"
+  "ArrayRef<Value> lenParams = {}, ArrayRef<Value> sizes = {},"
+  "ArrayRef<NamedAttribute> attributes = {}",
+  [{
+    result.addTypes(getRefTy(inType));
+    result.addAttribute("in_type", mlir::TypeAttr::get(inType));
+    result.addAttribute("name", builder->getStringAttr(name));
+    result.addOperands(sizes);
+    for (auto namedAttr : attributes)
+      result.addAttribute(namedAttr.first, namedAttr.second);
+  }]>;
+
+def fir_OneResultOpBuilder : OpBuilder<
+  "Builder *, OperationState &result, Type resultType,"
+  "ArrayRef<Value> operands, ArrayRef<NamedAttribute> attributes = {}",
+  [{
+    if (resultType)
+      result.addTypes(resultType);
+    result.addOperands(operands);
+    for (auto namedAttr : attributes)
+      result.addAttribute(namedAttr.first, namedAttr.second);
+  }]>;
+
+// Base class of FIR operations that return 1 result
+class fir_OneResultOp<string mnemonic, list<OpTrait> traits = []> :
+    fir_Op<mnemonic, traits>, Results<(outs fir_Type:$res)> {
+  let builders = [fir_OneResultOpBuilder];
+}
+
+// Base class of FIR operations that have 1 argument and return 1 result
+class fir_SimpleOneResultOp<string mnemonic, list<OpTrait> traits = []> :
+    fir_SimpleOp<mnemonic, traits> {
+  let builders = [fir_OneResultOpBuilder];
+}
+
+class fir_TwoBuilders<OpBuilder b1, OpBuilder b2> {
+  list<OpBuilder> builders = [b1, b2];
+}
+
+class fir_AllocatableBaseOp<string mnemonic, list<OpTrait> traits = []> :
+    fir_Op<mnemonic, traits>, Results<(outs fir_Type:$res)> {
+  let arguments = (ins
+    OptionalAttr<StrAttr>:$name,
+    OptionalAttr<BoolAttr>:$target
+  );
+}
+
+class fir_AllocatableOp<string mnemonic, list<OpTrait> traits =[]> :
+    fir_AllocatableBaseOp<mnemonic, !listconcat([NoSideEffect], traits)>,
+    fir_TwoBuilders<fir_AllocateOpBuilder, fir_NamedAllocateOpBuilder>,
+    Arguments<(ins TypeAttr:$in_type, Variadic<AnyIntegerType>:$args)> {
+
+  let parser = [{
+    mlir::Type intype;
```

Comment:
Could you please move all of these large code blocks to a .cpp source file? It is much easier to reason about/format/refactor code when it is in a .cpp file. It also makes the .td file much cleaner.

Reply:
Right. I'm wondering if such a NFC might be rolled up in another merge though to reduce any noise. (This is already a large PR.)

Reply:
@jeanPerier If this request is being deferred for now, please add it to the list of tasks for the FIR integration.

Reply:
We'll move some code around in a subsequent PR.

Reply:
Added on list of task.